File: src/main/java/org/example/eventhub/service/EventService.java
package org.example.eventhub.service;

import lombok.RequiredArgsConstructor;
import org.example.eventhub.dto.event.EventUpdateRequest;
import org.example.eventhub.exception.EventNotFoundException;
import org.example.eventhub.dto.event.EventCreateRequest;
import org.example.eventhub.dto.event.EventResponseLong;
import org.example.eventhub.dto.event.EventResponseShort;
import org.example.eventhub.entity.Event;
import org.example.eventhub.mapper.EventMapper;
import org.example.eventhub.repository.EventRepository;
import org.springframework.stereotype.Service;

import org.example.eventhub.entity.User;


import java.util.List;

@Service
@RequiredArgsConstructor
public class EventService {

    private final EventRepository repository;
    private final UserService userService;
    private final EventMapper mapper;

    public EventResponseLong createEvent(EventCreateRequest dto, Long organizerId) { // TODO возможно проверку на существование
        User user = userService.getUserByIdAsEntity(organizerId);
        Event event = mapper.toEntity(dto, user);
        Event saved = repository.save(event);

        return mapper.toLongDto(saved);
    }

    public EventResponseLong getEventById(Long id) {
        Event event = repository.findById(id)
                .orElseThrow(() -> new EventNotFoundException("Event с id" + id + " не найдено"));

        return mapper.toLongDto(event);
    }

    public Event getEventByIdAsEntity(Long id) {
        return repository.findById(id).orElseThrow(() -> new EventNotFoundException("Event с id " + id + " не найден"));
    }

    public List<EventResponseShort> getAllEvents() { // TODO pageable
        return repository.findAll()
                .stream()
                .map(mapper::toShortDto)
                .toList();
    }

    public EventResponseLong updateEvent(Long id, EventUpdateRequest dto) {
        Event event = repository.findById(id)
                .orElseThrow(() -> new EventNotFoundException("Event с id" + id + " не найдено"));

        if (dto.title() != null) event.setTitle(dto.title());
        if (dto.description() != null) event.setDescription(dto.description());
        if (dto.dateTime() != null) event.setDateTime(dto.dateTime());
        if (dto.location() != null) event.setLocation(dto.location());
        if (dto.capacity() != null) event.setCapacity(dto.capacity());
        if (dto.price() != null) event.setPrice(dto.price());
        if (dto.eventStatus() != null) event.setEventStatus(dto.eventStatus());

        return mapper.toLongDto(repository.save(event));
    }

    public void deleteEvent(Long id) {
        Event event = repository.findById(id)
                .orElseThrow(() -> new EventNotFoundException("Event с id" + id + " не найдено"));

        repository.deleteById(id);
    }

//    public List<EventResponseShort> getEventsByFilter(
//            String title, String city, LocalDateTime dateTime, Integer price) TODO
}

File: src/main/kotlin/org/example/eventhub/service/EventService.kt
package org.example.eventhub.service

import org.example.eventhub.dto.event.*
import org.example.eventhub.entity.Event
import org.example.eventhub.exception.EventNotFoundException
import org.example.eventhub.exception.NoAccessException
import org.example.eventhub.mapper.EventMapper
import org.example.eventhub.repository.EventRepository
import org.example.eventhub.specification.EventSpecification
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class EventService(
    private val repository: EventRepository,
    private val userService: UserService,
    private val mapper: EventMapper,
    private val specification: EventSpecification
) {

    fun createEvent(dto: EventCreateRequest, userDetails: UserDetails): EventResponseLong {
        val user = userService.getUserByEmailAsEntity(userDetails.username)
        val event = mapper.toEntity(dto, user)
        return mapper.toLongDto(repository.save(event))
    }

    fun getEventById(eventId: Long): EventResponseLong =
        mapper.toLongDto(getEventByIdAsEntity(eventId))

    fun getAllEvents(pageable: Pageable, eventFilter: EventFilter): Page<EventResponseShort> =
        repository.findAll(specification.withFilter(eventFilter), pageable)
            .map(mapper::toShortDto)

    fun updateEvent(userDetails: UserDetails, eventId: Long, dto: EventUpdateRequest): EventResponseLong {
        val event = getEventByIdAsEntity(eventId)

        checkAccess(userDetails, event)

        mapper.updateEntity(dto, event)
        return mapper.toLongDto(event)
    }

    fun deleteEvent(userDetails: UserDetails, eventId: Long) {
        val event = getEventByIdAsEntity(eventId)

        checkAccess(userDetails, event)

        repository.deleteById(eventId)
    }

    fun getEventByIdAsEntity(eventId: Long): Event =
        repository.findById(eventId)
            .orElseThrow { EventNotFoundException("Event с id $eventId не найден") }

    fun saveEvent(event: Event) =
        repository.save(event)

    private fun checkAccess(userDetails: UserDetails, event: Event) {
        val user = userService.getUserByEmailAsEntity(userDetails.username)

        if (event.organizer != user)
            throw NoAccessException(
                "Вы не можете изменять мероприятие ${event.title}, так как не являетесь его организатором"
            )
    }
}
